modules/network/src/main/java/org/apache/ignite/internal/network/recovery/RecoveryClientHandshakeManager.java

```diff
@@ -136,14 +158,35 @@ public void onMessage(NetworkMessage message) {
             this.remoteLaunchId = msg.launchId();
             this.remoteConsistentId = msg.consistentId();
 
-            this.recoveryDescriptor = recoveryDescriptorProvider.getRecoveryDescriptor(
+            RecoveryDescriptor descriptor = recoveryDescriptorProvider.getRecoveryDescriptor(
                     remoteConsistentId,
                     remoteLaunchId,
-                    connectionId,
-                    false
+                    connectionId
             );
 
-            handshake(recoveryDescriptor);
+            while (!descriptor.acquire(ctx)) {
+                if (launchId.compareTo(remoteLaunchId) > 0) {
```

Comment:
Please add clarifying comment, that this is a tie breaker in case of simultaneous connection openings from both side. It's not obvious from the code

Reply:
Sure, I forgor

Reply:
Considering that this code is duplicated 2 times, I strongly suggest extracting it into a single place.